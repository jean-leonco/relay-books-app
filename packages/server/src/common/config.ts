import path from 'path';

import dotEnvSafe from 'dotenv-safe';
import envVar from 'env-var';

import { version } from '../../package.json';

export const PACKAGE_VERSION = version;

const cwd = process.cwd();

const root = path.join.bind(cwd);

dotEnvSafe.config({
  // allowEmptyValues: process.env.NODE_ENV !== 'production',
  allowEmptyValues: true,
  path: root('.env'),
  sample: root('.env.example'),
});

// expose BOOKSAPP_ENV
export const BOOKSAPP_ENV = envVar.get('BOOKSAPP_ENV').required().asString();

// expose COMMIT_SHA
export const COMMIT_SHA = envVar.get('COMMIT_SHA').required().asString();

// Export some settings that should always be defined
export const MONGO_URI = envVar.get('MONGO_URI').required().asString();

export const JWT_KEY = envVar.get('JWT_KEY').required().asString();

// Host
export const GRAPHQL_HOST = envVar.get('GRAPHQL_HOST').asString();

// Ports
export const GRAPHQL_PORT = envVar.get('GRAPHQL_PORT').default('5001').asPortNumber();

export const DEBUG_GRAPHQL = envVar.get('DEBUG_GRAPHQL').default('false').asBoolStrict();
