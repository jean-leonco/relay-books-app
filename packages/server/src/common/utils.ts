import { Model, Types } from 'mongoose';
import { fromGlobalId } from 'graphql-relay';

// https://stackoverflow.com/a/2593661/710693
export const escapeRegex = (str: string) => `${str}`.replace(/[.?*+^$[\]\\(){}|-]/g, '\\$&');

export const PLATFORM = {
  APP: 'APP',
  UNKNOWN: 'UNKNOWN',
};

export const getPlatform = (platform?: string | null): string => {
  if (typeof platform !== 'string') {
    return PLATFORM.UNKNOWN;
  }

  //if (platform.includes('ios') || platform.includes('android')) {
  //  return PLATFORM.APP;
  //}

  if (platform === PLATFORM.APP) {
    return PLATFORM.APP;
  }

  return PLATFORM.UNKNOWN;
};

export function kilometersToEarthRadians(kilometers: number) {
  return kilometers / 6378.1;
}

// returns an ObjectId given an param of unknown type
export const getObjectId = (target: string | Model<any> | Types.ObjectId | undefined | null): Types.ObjectId | null => {
  if (!target) {
    return null;
  }

  if (target instanceof Types.ObjectId) {
    return new Types.ObjectId(target.toString());
  }

  if (typeof target === 'object') {
    return target && target._id ? new Types.ObjectId(target._id) : null;
  }

  if (typeof target === 'string') {
    const result = fromGlobalId(target);

    if (result.type && result.id && Types.ObjectId.isValid(result.id)) {
      return new Types.ObjectId(result.id);
    }

    if (Types.ObjectId.isValid(target)) {
      return new Types.ObjectId(target);
    }

    return null;
  }

  return null;
};

type EnumObject = {
  [key: string]: string;
};

type EnumObjectResult = {
  [key: string]: {
    value: string;
  };
};

export const enumBuilderValues = (constants: EnumObject): EnumObjectResult =>
  Object.keys(constants).reduce(
    (prev, curr) => ({
      ...prev,
      [curr]: {
        value: constants[curr],
      },
    }),
    {},
  );
