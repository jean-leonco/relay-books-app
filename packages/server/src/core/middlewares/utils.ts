import { GraphQLResolveInfo } from 'graphql';
import { ObjectSchema, Shape } from 'yup';

import { GraphQLContext } from '../../types';

interface GraphQLMutationConfig {
  validationSchema?(context: GraphQLContext): ObjectSchema<Shape<any, any>>;
  authenticatedOnly?: boolean;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): GraphQLMutationConfig | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
