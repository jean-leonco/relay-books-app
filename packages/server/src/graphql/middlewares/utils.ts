import { GraphQLField, GraphQLResolveInfo } from 'graphql';

import { Extensions } from '../../types';

interface Field extends GraphQLField<any, any> {
  extensions: Extensions | null | undefined;
}

export function getMutationFields(info: GraphQLResolveInfo, mutationName: string): Field | null {
  const mutationType = info.schema.getMutationType();
  if (!mutationType) {
    return null;
  }

  const fields = mutationType.getFields();
  return fields[mutationName];
}
