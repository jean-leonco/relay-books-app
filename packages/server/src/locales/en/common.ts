const common = {
  SomethingWentWrong: 'Something went wrong.',
  TheDescriptionShouldNot: 'The description must be less than {{size}} characters.',
};
export default common;
export type Keys = keyof typeof common;
