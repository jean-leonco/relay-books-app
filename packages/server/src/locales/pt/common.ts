const common = {
  SomethingWentWrong: 'Algo deu errado.',
  TheDescriptionShouldNot: 'A descrição deve ter menos de {{size}} caracteres.',
};
export default common;
export type Keys = keyof typeof common;
