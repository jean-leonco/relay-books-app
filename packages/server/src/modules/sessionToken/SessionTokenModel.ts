import mongoose, { Document, Model, Types } from 'mongoose';

import { PLATFORM } from '../../common/utils';
import { isActiveMongooseField } from '../../core/mongoose/withMongooseFields';

const { ObjectId } = mongoose.Schema.Types;

export enum SESSION_TOKEN_SCOPES {
  RESET_PASSWORD = 'BOOKSAPP:RESET_PASSWORD', // used when user requested a password reset
  SESSION = 'BOOKSAPP:SESSION', // token for a valid user session
}

const Schema = new mongoose.Schema(
  {
    user: {
      type: ObjectId,
      ref: 'User',
      description: 'User owner of this token.',
    },
    ip: {
      type: String,
      description: 'Agent IP that came from request headers.',
      required: true,
      index: true,
    },
    channel: {
      type: String,
      required: true,
      default: PLATFORM.APP,
      description: 'Channel that this token belongs too. ex: APP',
      enum: Object.values(PLATFORM),
    },
    name: {
      type: String,
      description: 'Name of this token. ex: Jean Linux',
    },
    scope: {
      type: String,
      description: 'Scope of this session token. ex: SESSION',
    },
    expiresIn: {
      type: Date,
      description: 'When this session token expires.',
    },
    isBlocked: {
      type: Boolean,
      description: 'If the token is blocked.',
      default: false,
      required: true,
      index: true,
    },
    ...isActiveMongooseField,
  },
  {
    timestamps: {
      createdAt: 'createdAt',
      updatedAt: 'updatedAt',
    },
    collection: 'SessionToken',
  },
);

export interface ISessionToken extends Document {
  user: Types.ObjectId;
  ip: string;
  channel: string;
  isActive: boolean;
  isBlocked: boolean;
  name: string;
  scope: string;
  expiresIn: Date | null;
  createdAt: Date;
  updatedAt: Date;
}

const SessionTokenModel: Model<ISessionToken> = mongoose.model('SessionToken', Schema);

export default SessionTokenModel;
