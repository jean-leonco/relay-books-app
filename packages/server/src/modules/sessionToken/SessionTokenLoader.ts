import DataLoader from 'dataloader';
import { Types } from 'mongoose';
import { isPast } from 'date-fns';

import { connectionFromMongoCursor, mongooseLoader } from '@entria/graphql-mongoose-loader';

import { GraphQLContext } from '../../types';

import { IUser } from '../user/UserModel';

import SessionTokenModel, { ISessionToken } from './SessionTokenModel';

export default class SessionToken {
  public registeredType = 'SessionToken';

  id: string;
  _id: Types.ObjectId;
  user: Types.ObjectId;
  ip: string;
  channel: string;
  scope: string;
  isActive: boolean;
  isBlocked: boolean;
  name: string;
  createdAt: Date;
  updatedAt: Date;
  expiresIn: Date | null;

  constructor(data: ISessionToken) {
    this.id = data.id || data._id;
    this._id = data._id;
    this.user = data.user;
    this.ip = data.ip;
    this.channel = data.channel;
    this.scope = data.scope;
    this.isActive = data.isActive;
    this.isBlocked = data.isBlocked;
    this.name = data.name;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
    this.expiresIn = data.expiresIn;
  }
}

export const getLoader = () => new DataLoader(ids => mongooseLoader(SessionTokenModel, ids));

// @TODO - improve this
const viewerCanSee = () => true;

export const load = async (context: GraphQLContext, id: string) => {
  if (!id) {
    return null;
  }

  try {
    const data = await context.dataloaders.SessionTokenLoader.load(id.toString());

    if (!data) {
      return null;
    }

    return viewerCanSee() ? new SessionToken(data) : null;
  } catch (err) {
    return null;
  }
};

export const clearCache = ({ dataloaders }: GraphQLContext, id: string) =>
  dataloaders.SessionTokenLoader.clear(id.toString());

// @TODO - do not query with isBlocked so we can validate if exists a session token
// blocked for this user and abort token generation process
export const loadValidSessionToken = async (
  ctx: GraphQLContext,
  user: IUser,
  channel: string,
  scope: string,
): Promise<ISessionToken | null> => {
  const { koaContext } = ctx;

  const { ip } = koaContext.request;
  const session = await SessionTokenModel.findOne({
    user: user._id,
    ip,
    channel,
    scope,
    isActive: true,
    isBlocked: false,
  }).lean<ISessionToken>();

  if (!session || isPast(session.expiresIn!)) return null;

  return session;
};

type SessionTokenArgs = {};

export const loadAll = async (context: GraphQLContext, args: SessionTokenArgs) => {
  const sessionTokens = SessionTokenModel.find().sort({ createdAt: -1 });

  return connectionFromMongoCursor({
    cursor: sessionTokens,
    context,
    args,
    loader: load,
  });
};

export const loadSessionTokensByIp = async (context: GraphQLContext, args: SessionTokenArgs, ip: string) => {
  const sessionTokens = SessionTokenModel.find({ ip }).sort({ createdAt: -1 });

  return connectionFromMongoCursor({
    cursor: sessionTokens,
    context,
    args,
    loader: load,
  });
};

export const loadSessionTokensByUser = async (context: GraphQLContext, args: SessionTokenArgs, user: string) => {
  const sessionTokens = SessionTokenModel.find({ user }).sort({ createdAt: -1 });

  return connectionFromMongoCursor({
    cursor: sessionTokens,
    context,
    args,
    loader: load,
  });
};
