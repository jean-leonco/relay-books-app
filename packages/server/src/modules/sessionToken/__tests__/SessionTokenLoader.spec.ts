import { sanitizeTestObject } from '@booksapp/test-utils';

import {
  clearDbAndRestartCounters,
  connectMongoose,
  createSessionToken,
  createUser,
  disconnectMongoose,
  getContext,
} from '../../../../test/helpers';

import * as SessionTokenLoader from '../SessionTokenLoader';
import { PLATFORM } from '../../../common/utils';
import { SESSION_TOKEN_SCOPES } from '../SessionTokenModel';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

describe('SessionTokenLoader', () => {
  it('should load null when looking for valid session token', async () => {
    const user = await createUser();

    await createSessionToken({ user, isBlocked: true });

    const context = await getContext({ user });

    const validSessionToken = await SessionTokenLoader.loadValidSessionToken(
      context,
      user,
      PLATFORM.APP,
      SESSION_TOKEN_SCOPES.SESSION,
    );

    expect(validSessionToken).toBeNull();
  });

  it('should load a valid session token', async () => {
    const user = await createUser();

    await createSessionToken({ user: await createUser(), isBlocked: true });
    await createSessionToken({ user: await createUser(), isActive: false });
    await createSessionToken({ user });

    const context = await getContext({ user });

    const validSessionToken = await SessionTokenLoader.loadValidSessionToken(
      context,
      user,
      PLATFORM.APP,
      SESSION_TOKEN_SCOPES.SESSION,
    );

    expect(validSessionToken?.user).toEqual(user._id);
    expect(sanitizeTestObject(validSessionToken)).toMatchSnapshot();
  });
});
