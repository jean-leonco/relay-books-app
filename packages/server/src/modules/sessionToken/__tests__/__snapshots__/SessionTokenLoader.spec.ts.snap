// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SessionTokenLoader should load a valid session token 1`] = `
Object {
  "__v": 0,
  "_id": "ObjectId",
  "channel": "APP",
  "createdAt": "FROZEN-CREATEDAT",
  "ip": "::ffff:127.0.0.1",
  "isActive": true,
  "isBlocked": false,
  "scope": "BOOKSAPP:SESSION",
  "updatedAt": "FROZEN-UPDATEDAT",
  "user": "ObjectId",
}
`;
